Clean up single vs. double brace handling

gwicke opened the pull request:
- Consistently escape single brace expressions according to the rules of URI
  templates. This lets us sanely support general path segment templating as
  needed for https://phabricator.wikimedia.org/T126571.
- Remove old-style $ prefixes from tests, but still support them.

gwicke committed:
Support newlines in expressions & release v0.3.4

Template expressions are currently limited to a single line, which forces us
to squeeze complex expressions into long lines. These expressions would be a
lot more readable if we could break them up into multiple lines, and indent
them to reflect the logical structure.

This patch adds simple multi-line expression support by simply replacing
newlines with a space.

gwicke committed:
Clean up single vs. double brace handling

- Consistently escape single brace expressions according to the rules of URI
  templates. This lets us sanely support general path segment templating as
  needed for https://phabricator.wikimedia.org/T126571.
- Remove old-style $ prefixes from tests, but still support them.

gwicke commented:
/cc @wikimedia/services

gwicke committed:
Release v0.4.0

Also, remove a few outdated comments.

gwicke commented:
(rebased on master)

Pchelolo commented on lib/reqTemplate.js:
```diff
@@ -138,7 +140,7 @@ function splitAndPrepareTAsseblyTemplate(templateSpec, options) {
                 // TODO: Enable single brace structures outside of URI once
                 // the specs are migrated to double brace.
-                if (!inDoubleBrace && options.isURI) {
+                if (!inDoubleBrace) {
```
To the comment above: given that we have https://github.com/wikimedia/restbase/pull/516 and it's almost done, we can resolve that TODO right now and simultaneously merge RB PR.

gwicke replied:
Actually, this PR does that already. Let me remove the comment.

Pchelolo commented on lib/reqTemplate.js:
```diff
@@ -138,7 +140,7 @@ function splitAndPrepareTAsseblyTemplate(templateSpec, options) {
                 // the specs are migrated to double brace.
-                if (!inDoubleBrace && options.isURI) {
+                if (!inDoubleBrace) {
                     if (/^\+/.test(currentTemplate)) {
```
We can actually drop support for `+`, as it's equivalent to double braces.

gwicke replied:
Double braces are a custom extension and not part of the URI template spec. This means that we can't use them in public path templates. I think there is value in keeping path templates and request handler templates consistent, so I would recommend to keep `{+path}` working.

Pchelolo replied:
hm.. We don't really evaluate the templates over public uri specs now, but I'm ok with leaving support.

Pchelolo commented on lib/reqTemplate.js:
```diff
@@ -263,11 +264,11 @@ function replaceComplexTemplates(part, subSpec, globals) {
+        if (/\{[^\}]+\}/.test(subSpec)) {
             // There is a template, now we need to check it for special stuff we replace
-            if (/^\{[^\}]+\}$/.test(subSpec)) {
+            if (/^\{\{[^\{\}]+\}\}$/.test(subSpec)) {
```
We should also support single expressions with single braces. It's not that useful, but we need it for consistency. Actually, I think that it should work now within the `splitAndPrepareTAsseblyTemplate`, and actually this branch is useless.

gwicke replied:
I believe we introduced this branch as a performance optimization, but it seems unlikely that it makes a big difference to startup performance.

Pchelolo commented on test/features/reqTemplate.js:
```diff
@@ -468,6 +469,36 @@ describe('Request template', function() {
+            uri: '/a/host/a%2Ffoo/',
+            headers: {
+                bar: 'a/bar',
+                // FIXME: This will change in the future!
```
What will change in future?

gwicke replied:
Looks like a stray comment.

gwicke committed:
Remove special-case "simple expression" optimization

We used to have a fast path for simple expression compilation, but it is not
clear that the performance gains are worth the special-casing. This code only
runs at startup, and it is unlikely that splitAndPrepareTAsseblyTemplate is
significantly slower.

Also, fix the spelling of splitAndPrepareTAsseblyTemplate.

Pchelolo commented:
LGTM

Pchelolo commented on lib/reqTemplate.js:
```diff
@@ -236,7 +235,7 @@ function createURIResolver(uri, globals) {
         };
     } else if (/\{/.test(uri)) {
-        var tassemblyTemplate = splitAndPrepareTAsseblyTemplate(uri, { isURI: true });
+        var tassemblyTemplate = splitAndPrepareTAssemblyTemplate(uri, { isURI: true });
```
The `isURI` option is not used any more.

gwicke committed:
Minor: Remove no-longer needed isURI option

d00rman commented:
Published as [v0.4.0](https://github.com/wikimedia/swagger-router/tree/v0.4.0)
